amqp/src/main/java/io/github/tcdl/msb/adapters/amqp/AmqpMessageConsumer.java

```diff
@@ -41,27 +43,34 @@ public AmqpMessageConsumer(Channel channel, ExecutorService consumerThreadPool,
 
     @Override
     public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
+        long deliveryTag = envelope.getDeliveryTag();
+        AmqpAcknowledgementHandler ackHandler = createAcknowledgementHandler(
+                getChannel(), consumerTag, deliveryTag, amqpBrokerConfig.isRequeueRejectedMessages());
         try {
             Charset charset = amqpBrokerConfig.getCharset();
-            boolean requeueRejectedMessages = amqpBrokerConfig.isRequeueRejectedMessages();
+
             String bodyStr = new String(body, charset);
+
             LOG.debug(String.format("[consumer tag: %s] Message consumed from broker: %s", consumerTag, bodyStr));
+
             try {
-                consumerThreadPool.submit(new AmqpMessageProcessingTask(consumerTag, bodyStr, getChannel(), envelope.getDeliveryTag(), msgHandler));
-                LOG.debug(String.format("[consumer tag: %s] Message has been put in the processing queue: %s. About to send AMQP ack...",
+                consumerThreadPool.submit(new AmqpMessageProcessingTask(consumerTag, bodyStr, msgHandler, ackHandler));
+                LOG.debug(String.format("[consumer tag: %s] Message has been put in the processing queue: %s.",
                         consumerTag, bodyStr));
-
-                getChannel().basicAck(envelope.getDeliveryTag(), false);
-                LOG.debug(String.format("[consumer tag: %s] AMQP ack has been sent for message '%s'", consumerTag, bodyStr));
             } catch (Exception e) {
-                LOG.error(String.format("[consumer tag: %s] Couldn't put message in the processing queue: %s. About to send AMQP reject...",
+                LOG.error(String.format("[consumer tag: %s] Couldn't put message in the processing queue: %s.",
                         consumerTag, bodyStr), e);
-                getChannel().basicReject(envelope.getDeliveryTag(), requeueRejectedMessages);
-                LOG.error(String.format("[consumer tag: %s] AMQP reject has been sent for message: %s", consumerTag, bodyStr));
+                throw e;
             }
         } catch (Exception e) {
             // Catch all exceptions to prevent AMQP channel to be closed
-            LOG.error(String.format("[consumer tag: %s] Got exception while processing incoming message", consumerTag), e);
+            LOG.error(String.format("[consumer tag: %s] Got exception while processing incoming message. About to send AMQP reject...", consumerTag), e);
+            ackHandler.autoReject();
         }
     }
+    
+    AmqpAcknowledgementHandler createAcknowledgementHandler(Channel channel, String consumerTag, long deliveryTag, boolean isRequeueRejectedMessages) {
+        return new AmqpAcknowledgementHandler(channel, consumerTag, deliveryTag, isRequeueRejectedMessages);
```

Comment:
We might have a problem here. Because we create handler per consumer. Suppose `isRequeueRejectedMessages` is  `true`. But in some cases (for example when the message cannot be parsed) it should not be requeued. I see 2 ways to overcome this limitation:
- introduce another methods in the handler interface `dropMessage`
- invoke `confirmMessage` in the cases described above